### PR TITLE
Release test harness as an artifact

### PR DESCRIPTION
**Pull request by KengoTODA**
This pull-request is based on #97, please merge that pull-request first.

As described in #53, I want to publish test-harness artifact for plugin developers.
This pull-request adds several tasks to build jar, javadoc-jar and source-jar for test-harness artifact.

BTW, in #59 I tried to create an isolated submodule, but it failed because core depends on several detectors directly (= cannot solve cyclic reference between spotbugs-core, test-harness and detectors).

### CODE REVIEW

**Comment by henrik242** [comment]
@KengoTODA  It looks like the maven-publish commit is here as well, I guess that's for the other PR?

File: findbugs/build.gradle [hunk 1]
```diff
@@ -181,6 +181,40 @@ task annotationSourcesJar(type: Jar, dependsOn: classes) {
...
+  from testHarnessJavadoc.destinationDir
+}
+
+task testHarnessSourcesJar(type: Jar, dependsOn: testClasses) {
+  classifier = 'sources'
+  baseName 'test-harness'
+  from sourceSets.test.allSource
+  include 'edu/umd/cs/findbugs/AbstractIntegrationTest.java'
+  include 'org/sonar/plugins/findbugs/resource/SmapParser.java'
+  include 'edu/umd/cs/findbugs/test/**/*.java'
```
Comment (jsotuyod) [inline findbugs/build.gradle]: these 4 lines are repeated over and over... I'm particularly worried about the includes. We should refactor this by extracting all 3 ant matchers to an array in a variable and using that directly (`include` supports it). We may also want to start splitting this build file into one per artifact under `gradle` dir.

Something such as:

```
def harnessIncludes = ['edu/umd/cs/findbugs/AbstractIntegrationTest.class',
  'org/sonar/plugins/findbugs/resource/SmapParser.class',
  'edu/umd/cs/findbugs/test/**/*.class']

task testHarnessJar(type: Jar, dependsOn: testClasses) {
  baseName 'test-harness'
  from sourceSets.test.output
  include harnessIncludes
}

task testHarnessJavadoc(type: Javadoc) {
  source sourceSets.test.allSource
  include harnessIncludes
  destinationDir = file("${project.docsDir}/test-harness")
  options.with {
    tags = ['Test']
  }
}

task testHarnessJavadocJar(type: Jar, dependsOn: testHarnessJavadoc) {
  classifier = 'javadoc'
  baseName 'test-harness'
  from testHarnessJavadoc.destinationDir
}

task testHarnessSourcesJar(type: Jar, dependsOn: testClasses) {
  classifier = 'sources'
  baseName 'test-harness'
  from sourceSets.test.allSource
  include harnessIncludes
}

plugins.withType(org.gradle.api.publish.maven.plugins.MavenPublishPlugin) { // lazy as plugin may not applied yet
  publishing {
    publications {
      testHarness(MavenPublication) {
        artifactId 'test-harness'
        artifact testHarnessJar

        artifact testHarnessSourcesJar {
          classifier 'sources'
        }

        artifact testHarnessJavadocJar {
          classifier 'javadoc'
        }
      }
    }
  }
}
```

Also of note, this should probably be included in `artifacts { }` too, shouldn't it?

**Comment by KengoTODA** [comment]
Thanks @jsotuyod for your review! I've committed e920357 to reduce duplicated code, please check.

And `artifacts { }` should be fixed by 0c2388c.

**Comment by jsotuyod** [comment]
@KengoTODA Travis is failing. Javadoc generation under Java 9 throws an error.

**Comment by henrik242** [comment]
@KengoTODA Try rebasing this on top of the latest master, maybe we'll get a green build

**Pull request merged**